Content.Client/GameObjects/Components/Instruments/InstrumentComponent.cs

```diff
@@ -266,29 +286,71 @@ public bool CloseMidi()
         /// <param name="midiEvent">The received midi event</param>
         private void RendererOnMidiEvent(MidiEvent midiEvent)
         {
-            _midiQueue.Enqueue(midiEvent);
+            _midiQueue.Add(midiEvent);
         }
 
-        public override void Update(float delta)
-        {
-            if (!IsMidiOpen && !IsInputOpen)
-                return;
 
-            _timer -= delta;
+        private TimeSpan _lastMeasuredInterval = TimeSpan.MinValue;
 
-            if (_timer > 0f) return;
+        private int _sentWithinASec = 0;
 
-            SendAllMidiMessages();
-            _timer = TimeBetweenNetMessages;
-        }
+        private static readonly TimeSpan OneSecAgo = TimeSpan.FromSeconds(-1);
 
-        private void SendAllMidiMessages()
+        private readonly Dictionary<byte, TimeSpan> _notesActive = new Dictionary<byte,TimeSpan>();
+
+        public override void Update(float delta)
         {
+            if (!IsMidiOpen && !IsInputOpen) return;
+
+            var now = _gameTiming.RealTime;
+            var oneSecAGo = now.Add(OneSecAgo);
+
+            if (_lastMeasuredInterval <= oneSecAGo)
+            {
+                _lastMeasuredInterval = now;
+                _sentWithinASec = 0;
+            }
+
             if (_midiQueue.Count == 0) return;
-            var events = _midiQueue.ToArray();
-            _midiQueue.Clear();
+
+            var max = Math.Min(MaxMidiEventsPerBatch, MaxMidiEventsPerSecond - _sentWithinASec);
+
+            if (max <= 0)
+            {
+                _midiSawmill.Info($"Playing {_notesActive.Count} notes, hit {_sentWithinASec} event/sec limit.");
+                return;
+            }
+
+            var events = _midiQueue.Distinct().Take(max).ToArray();
+
+            foreach (var ev in events)
+            {
+                switch (ev.Type)
+                {
+                    case 0x80 /*NOTE_OFF*/:
+                    {
+                        _notesActive.Remove(ev.Key);
+                        break;
+                    }
+                    case 0x90 /*NOTE_ON*/:
+                    {
+                        _notesActive[ev.Key] = now;
+                        break;
```

Comment:
This is a terrible hack. Also NOTE_ON with velocity 0 do exactly the same as NOTE_OFF.

Reply:
Yeah it didn't do anything but try to count active notes. Good to know that there's other things that do the same thing as NOTE_OFF, would have been helpful to know that when I was working on it.